Add information about moving storage locations and improve lifecycle guide (#407)
filecoin-project/filecoin-docs: docs/mine/lotus/miner-lifecycle.md

---
title: 'Lotus Miner: lifecycle'
description: 'How to safely perform maintenance on a Lotus miner.'
breadcrumb: Lifecycle'
---

# {{ $frontmatter.title }}

{{ $frontmatter.description }}

These operations are normally related to maintenances and upgrades. Given that miners are expected to submit proofs to the chain in a continuous fashion while running long and expensive operations, it is important that operators become familiar with how to manage some of the events in the miner lifecycle so that they can be performed with the maximum guarantees.

[[TOC]]

## Safely restarting the miner daemon

> Before shutting down your miner daemon, make sure that there are no pending operations that could get your miner _slashed_.

Reviewer: ```suggestion
Before shutting down your miner daemon, make sure that there are no pending operations that could get your miner _slashed_ if they are delayed or cancelled.
```

Reviewer 2: This sentence should point to a section on how to check the mpool for pending messages.

Author: Isn't enough with the `proving info` below (or `proving dearlines`)?

Author: > how to check the mpool for pending messages.

Can you actually tell me how?

Author: Another question is how to make a miner stop taking deals without restarting it (so that we can be sure it has not just closed another deal and it's receiving data when its stopped).

Reviewer 2: Actually, yes to restart just the miner checking proving is sufficient. But in case here is how you check pending messages, `lotus mpool pending --local | jq -c . | wc -l` will give you the number of pending messages. To stop taking deals you can do `lotus-miner storage-deals selection reject --online --offline` and `lotus-miner retrieval-deals selection reject --online --offline`

> Before shutting down your miner daemon, make sure that there are no pending operations that could get your miner _slashed_.

Reviewer: It would be nice to have some expectation-setting like "A simple restart usually takes about X minutes" and any other caveats around time.

Reviewer 2: A restart will take as long as it takes for your machine to verify the proof params

Reviewer 2: As such it would be a good suggestion that miners put the params onto nvme drives so that they can be quickly read.

Author: added this suggestion to the setup instructions.

```bash
$ sudo lotus-miner proving info

Miner: t01000
Current Epoch:           497
Proving Period Boundary: 658
Proving Period Start:    658 (in 1h20m30s)
Next Period Start:       3538 (in 25h20m30s)
Faults:      0 (0.00%)
Recovering:  0
Deadline Index:       0
Deadline Sectors:     768
Deadline Open:        658 (in 1h20m30s)
Deadline Close:       718 (in 1h50m30s)
Deadline Challenge:   638 (in 1h10m30s)
Deadline FaultCutoff: 588 (in 45m30s)
```

Once you have verified that your miner is safe to shut down, you can run `lotus-miner stop`. If you are using systemd, run `systemctl stop lotus-miner`.

You can restart the miner as soon as you wish. Any sealing operations that were ongoing will be reset to the last checkpoint and continue.

Workers do not need to be restarted as they will reconnect to the miner automatically when it comes back up. However, if you are upgrading your node at the same time as shutting down your miner, you will need to restart the machine.

## Restarting workers

Lotus [seal workers](seal-workers.md) can be restarted any time, but if they are in the middle of one of the sealing steps, then the operation will start again (from the last checkpoint).

> ::: warning
> There is a maximum of three attempts to complete a sealing operation before it is fully started from scratch (_pre-commit1_ phase).

Reviewer 2: ```suggestion
::: warning
There is a maximum of three attempts to complete the _precommit2_ operation before sealing is fully started from scratch (_pre-commit1_ phase).
```

:::

## Changing storage locations

If you wish to change the location of your miner-related storage to a different path, for the miner or the seal workers, make sure that the Lotus miner and any seal workers are aware of the new location.

```sh
lotus-miner storage list
```

> The above command will give you an overview of [storage locations known to the miner](custom-storage-layout.md). This information is stored in `~/lotusminer/storage.json` (or `$LOTUS_MINER_PATH/storage.json` if defined). Lotus seal workers store all the data in the `~/lotusworker` folder (or `$LOTUS_WORKER_PATH` if defined).

Reviewer 2: ```suggestion
The above command will give you an overview of [storage locations known to the miner](custom-storage-layout.md). This information is stored in `~/.lotusminer/storage.json` (or `$LOTUS_MINER_PATH/storage.json` if defined). Lotus seal workers store all the data in the `~/.lotusworker` folder (or `$LOTUS_WORKER_PATH` if defined).
```

If you wish to change any of the storage locations of the Lotus miner, follow these steps:

> 1. [Stop the miner](#safely-restarting-the-miner-daemon)
> 2. Copy or move the data. You should move the data folder from the original location to the destination without changing any of the content.

Reviewer 2: ```suggestion
1. Copy or move the data. You should move the data folder from the original location to the destination without changing any of the content.
2. [Stop the miner](#safely-restarting-the-miner-daemon)
```
Stopping the miner first will result in definitely missing many window posts purely due to the amount of time it will take to transfer any non-trivial number of sectors. i.e. 1GbE LAN takes just over ~1hr to transfer 448GiBs (or 14 sectors).

Author: I'm guessing the data should not be moved then, while the miner is running, only copied. And I am guessing that even if the miner is running, it should not take on any new deals etc while the copy is ongoing?

Reviewer 2: > I'm guessing the data should not be moved then, while the miner is running, only copied

correct
> it should not take on any new deals etc while the copy is ongoing?

correct, also don't start pledging any new sectors, obviously.

If you wish to change the storage location for any of the lotus workers:

1. Stop the Lotus Worker.
2. Move the data to the new location.
3. Set `$LOTUS_WORKER_PATH` accordingly.
4. Start the worker again.

Any operations that the worker was performing before stopping will be restarted from the last checkpoint.

:::warning
Moving data between different workers is not currently supported. Moving the worker storage folder to a different worker machine will not work as the miner expects the ongoing sealing operations to be completed by the worker they were assigned to in the first place.
:::

## Using a different Lotus Node

If you are planning to run maintenance on the Lotus node used by the miner, or if you need to failover to a different Lotus node because the current one does not work, follow these steps:

1. [Stop the miner](#safely-restarting-the-miner-daemon)
2. Set the `FULLNODE_API_INFO` environment variable to the location of the new node:

```sh
export FULLNODE_API_INFO=<api_token>:/ip4/<lotus_daemon_ip>/tcp/<lotus_daemon_port>/http
```

Follow these steps to learn [how to obtain a token](../../build/lotus/api-token-generation.md).

> 3. If you have not exported your wallets yet, export them now from the old node and re-import them to the new Lotus node.

Reviewer 2: I don't think lotus currently supports offline importing of keys, but I could be mistaken.

Author: This assumes both old and new node are running. Export/import instructions also mention just copying the `keystore` folder though, I'll make it clearer.
